rfcs/0000-api-cr-crowding.md

```diff
@@ -0,0 +1,337 @@
+- Feature Name: api_cr_crowding
+- Start Date: 2021-09-23
+- RFC PR: [mbta/technology-docs#0000](https://github.com/mbta/technology-docs/pull/0000)
+- Asana task: [asana link](https://app.asana.com/)
+- Status: Proposed
+
+# Summary
+
+[summary]: #summary
+
+We want to provide predicted commuter rail crowding data for upcoming trips. The data will be
+consumed from a feed provided by Keolis. We will follow the lead of this [static GTFS-Occupanices
+proposal](https://github.com/google/transit/pull/240), but for now the data will only appear in
+the API and not in our GTFS feed. The API will gain a new "occupancies" resource. The data will be
+documented in the API's swagger docs as "Experimental" with a warning that it could change in the
+future. The implementation will be as minimal and simple as possible, but nevertheless built in a
```

Comment:
I think that because dotcom is internal to our organization we at least have the potential to tweak things (either by asking @thecristen very nicely and offering her a tasty coffee drink, or issuing the PR ourselves). So I'd like to reserve the right to make changes to the endpoint and not have to maintain previous versions. For example, we're passing along the `percentage` field for now, at Keolis's request and per their suggestion in the GTFS proposal. But if it doesn't make it and dotcom uses a design that doesn't expose that value, I'd like to be able to just remove that field from our data and response.

Versioning is possible, but adds overhead. Plus, there's the additional complication that dotcom isn't on the latest version now anyway, so they wouldn't really be able to easily update to use a newer variant of this endpoint.

Reply:
A thought: clients should have to explicitly opt into behavior that might have unannounced breaking changes. I think the most flexible way to do this would be via an HTTP header, which should also be familiar for clients since they already must do this to provide their API key. For example: `x-enable-experimental-features: true`. A plug could pick this up and put a boolean value into `conn.assigns[:experimental?]` or something.

This should work regardless of whether the data is provided as a standalone endpoint or a relationship. It would also provide a reusable framework for doing other "experimental" things with the API.

Reply:
I like that idea!